### PR TITLE
Allow to set timeout for jenkins pipelines

### LINKED ISSUE #2434: Add timeout to jenkins pipline
### Is your feature request related to a problem?
Even tho AJTS offers measures that prevent endless loop it could still happen(maybe somebody forgets the annotations). When a test runs for like 30 minutes it should be killed by Jenkins. Here is how it could be done:
https://stackoverflow.com/questions/38096004/how-to-add-a-timeout-step-to-jenkins-pipeline#38106655

### PR DESCRIPTION
**Pull request by axel1200**
Fixes #2434

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls

### Motivation and Context
Timeout out removes stale builds so i don't have to to it manually

### Steps for Testing

Use TS2 for testing
1. Create a programming exercise
2. Go to Edit in Editor, change to the Test Repository and remove the @StrictTimeout annotations from SortingBehaviorTest.java
3. Participate in the exercise
4. Submit valid code with an endless loop
5. Check the build plan in Jenkins. It should be stopped after 30 mins.

I have no set up server. So I would be cool if one could help me test this or something similar.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/DefaultJenkinsBuildPlanCreator.java [hunk 1]
```diff
@@ -1,33 +0,0 @@
...
-        super(resourceLoaderService, jenkinsService);
-    }
-
-    @Override
-    public String getPipelineScript(ProgrammingLanguage programmingLanguage, URL testRepositoryURL, URL assignmentRepositoryURL, boolean isStaticCodeAnalysisEnabled) {
-        final var resourcePath = new String[] { "templates", "jenkins", programmingLanguage.name().toLowerCase(), "Jenkinsfile" };
-
-        final var replacements = Map.of(REPLACE_TEST_REPO, testRepositoryURL.toString(), REPLACE_ASSIGNMENT_REPO, assignmentRepositoryURL.toString(), REPLACE_GIT_CREDENTIALS,
-                gitCredentialsKey, REPLACE_ASSIGNMENT_CHECKOUT_PATH, Constants.ASSIGNMENT_CHECKOUT_PATH, REPLACE_TESTS_CHECKOUT_PATH, Constants.TESTS_CHECKOUT_PATH,
-                REPLACE_ARTEMIS_NOTIFICATION_URL, artemisNotificationUrl, REPLACE_NOTIFICATIONS_TOKEN, ARTEMIS_AUTHENTICATION_TOKEN_KEY, REPLACE_DOCKER_IMAGE_NAME,
```
Comment (axel1200) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/DefaultJenkinsBuildPlanCreator.java:28]: That code was duplicated 3 time and it was not readable with thing was replaced by what

File: src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JavaJenkinsBuildPlanCreator.java [hunk 2]
```diff
@@ -23,25 +22,19 @@ public JavaJenkinsBuildPlanCreator(ResourceLoaderService resourceLoaderService,
...
+    @Override
+    protected Map<String, String> getReplacements(ProgrammingLanguage programmingLanguage, URL testRepositoryURL, URL assignmentRepositoryURL, boolean isStaticCodeAnalysisEnabled) {
+        Map<String, String> replacements = super.getReplacements(programmingLanguage, testRepositoryURL, assignmentRepositoryURL, isStaticCodeAnalysisEnabled);
+        if(isStaticCodeAnalysisEnabled){
             String staticCodeAnalysisScript = createStaticCodeAnalysisScript();
-            replacements = Map.of(REPLACE_TEST_REPO, testRepositoryURL.toString(), REPLACE_ASSIGNMENT_REPO, assignmentRepositoryURL.toString(), REPLACE_GIT_CREDENTIALS,
-                    gitCredentialsKey, REPLACE_ASSIGNMENT_CHECKOUT_PATH, Constants.ASSIGNMENT_CHECKOUT_PATH, REPLACE_ARTEMIS_NOTIFICATION_URL, artemisNotificationUrl,
-                    REPLACE_NOTIFICATIONS_TOKEN, ARTEMIS_AUTHENTICATION_TOKEN_KEY, REPLACE_STATIC_CODE_ANALYSIS_SCRIPT, staticCodeAnalysisScript, REPLACE_DOCKER_IMAGE_NAME,
-                    jenkinsService.getDockerImageName(ProgrammingLanguage.JAVA));
+            replacements.put(REPLACE_STATIC_CODE_ANALYSIS_SCRIPT, staticCodeAnalysisScript);
```
Comment (axel1200) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JavaJenkinsBuildPlanCreator.java:35]: Only set what is really different here

**Review by kloessst — COMMENTED** [review]
Code looks fine to me. Still have to test this though

**Comment by axel1200** [comment]
No idea why tests are failing

**Comment by kloessst** [comment]
You can run .\gradlew spotlessApply to automatically fix the style issues.

**Review by kloessst — CHANGES_REQUESTED** [review]
Can you quickly fix the codacy issues? I think both a valid. Double-brace initialization causes another .class file to be created.

**Review by fde312 — APPROVED** [review]
Tested it and the build was aborted after 30 mins:
image

As far as I can tell the code looks also good! Good job 🚀

**Review by kloessst — APPROVED** [review]
Worked for me. Jenkins stopped the build after 30 mins:

image

You have to remove the StrictTimeout annotations, otherwise AJTS will stop the endless loop (Good to know that this works 😃 )

This is the result. I'm not sure if we can improve this. I guess it is fine for now:
image

**Comment by axel1200** [comment]
Nice

**Review by krusche — APPROVED** [review]

**Pull request merged**